libkbfs/md_util.go

```diff
@@ -18,6 +19,57 @@ type mdRange struct {
 	end   MetadataRevision
 }
 
+func makeRekeyReadErrorHelper(
+	kmd KeyMetadata, resolvedHandle *TlfHandle, keyGen KeyGen,
+	uid keybase1.UID, username libkb.NormalizedUsername) error {
+	if resolvedHandle.IsPublic() {
+		panic("makeRekeyReadError called on public folder")
+	}
+	// If the user is not a legitimate reader of the folder, this is a
+	// normal read access error.
+	if !resolvedHandle.IsReader(uid) {
+		return NewReadAccessError(resolvedHandle, username)
+	}
+
+	// Otherwise, this folder needs to be rekeyed for this device.
+	tlfName := resolvedHandle.GetCanonicalName()
+	if hasKeys := kmd.HasKeyForUser(keyGen, uid); hasKeys {
+		return NeedSelfRekeyError{tlfName}
+	}
+	return NeedOtherRekeyError{tlfName}
+}
+
+func makeRekeyReadError(
+	ctx context.Context, config Config, kmd KeyMetadata, keyGen KeyGen,
+	uid keybase1.UID, username libkb.NormalizedUsername) error {
+	h := kmd.GetTlfHandle()
+	resolvedHandle, err := h.ResolveAgain(ctx, config.KBPKI())
+	if err != nil {
+		// Ignore error and pretend h is already fully
+		// resolved.
+		resolvedHandle = h
+	}
+	return makeRekeyReadErrorHelper(
+		kmd, resolvedHandle, keyGen, uid, username)
+}
+
+// Helper which returns nil if the md block is uninitialized or readable by
+// the current user. Otherwise an appropriate read access error is returned.
+func isReadableOrError(
+	ctx context.Context, config Config, md ReadOnlyRootMetadata) error {
```

Comment:
My reason for leaving it a free function is that it pulls in `Config`, and I prefer to not make the RMD types depend on `Config` whenever possible, since Config pulls in ~everything.